designs/NuGetExtensibility/INuGetProjectServices.md

```diff
@@ -0,0 +1,138 @@
+    public GetInstalledPackageResultStatus Status { get; }
+
+    /// <summary>List of packages in the project</summary>
+    /// <remarks>May be null if <see cref="Status"/> was not successful</remarks>
+    public IReadOnlyCollection<NuGetInstalledPackage> Packages { get; }
+
+    // To maximize the chance this class can be converted to a C#9 record type without breaking changes, constructor is not public
+    internal GetInstalledPackagesResult();
+}
+
+/// <summary>The status of the result</summary>
+public enum GetInstalledPackageResultStatus
+{
+    /// <summary>Unknown status</summary>
+    /// <remarks>Probably represents a bug in the method that created the result.</remarks>
+    Unknown = 0,
+
+    /// <summary>Successful</summary>
+    Successful,
+
+    /// <summary>The project is not yet ready</summary>
+    /// <remarks>This typically happens shortly after the project is loaded, but the project system has not yet informed NuGet about package references</remarks>
+    ProjectNotReady,
+
+    /// <summary>Package information could not be retrieved because the project is in an invalid state</summary>
+    /// <remarks>If a project has an invalid target framework value, or a package reference has a version value, NuGet may be unable to generate basic project information, such as requested packages.</remarks>
+    ProjectInvalid
+}
+
+/// <summary>Basic information about a package</summary>
+public sealed class NuGetInstalledPackage
+{
+    /// <summary>The package id.</summary>
+    public string Id { get; }
+
+    /// <summary>The package version</summary>
+    /// <remarks>
+    /// If the project uses packages.config, this will be the installed package version.
+    /// If the project uses PackageReference, this would ideally be the resolved package version, but may be the requested package version.
+    /// </remarks>
+    public string Version { get; }
+
+    // To maximize the chance this class can be converted to a C#9 record type without breaking changes, constructor is not public
+    internal NuGetInstalledPackage();
+}
+
+/// <summary>Factory to create types</summary>
+/// <remarks>Trying to be forwards compatible with what C#9 records are going to be</remarks>
+public static class ContractsFactory
+{
+    /// <summary>Create a <see cref="NuGetInstalledPackage"/></summary>
+    /// <param name="id"></param>
+    /// <param name="version"></param>
+    /// <param name="isDirectDependency"></param>
+    /// <returns></returns>
+    public static NuGetInstalledPackage CreateNuGetInstalledPackage(string id, string version, bool isDirectDependency);
+
+    /// <summary>Create a <see cref="GetInstalledPackageResultStatus"/></summary>
+    /// <param name="status"></param>
+    /// <param name="packages"></param>
+    /// <returns></returns>
+    public static GetInstalledPackagesResult CreateGetInstalledPackagesResult(GetInstalledPackageResultStatus status, IReadOnlyCollection<NuGetInstalledPackage> packages);
+}
```

Comment:
While looking forward to C# 9 record types is admirable, requiring a factory to create what is a simple serializable data type is _not_ API friendly, nor is it in anyway consistent with our other remotable service interface data types defined elsewhere. 
We do use immutable data types, but we define a public constructor with a parameter for each property. I am not sure, but I suspect records may offer similar public API. But even if it doesn't, a public constructor instead of a factory method on some other random type is very difficult to discover.

Reply:
I understand and agree with your comment. I think discoverability can be addressed to some extent with XMLdoc that points to the factory (see new commit once I push it). This might be insufficient for types that API customers need to create. But at this time, the two classes defined are both return-value only. Meaning the most likely reason customers might use it is for tests, assuming they don't have their own abstraction of our result (Roslyn does, so is unlikely to create instances).  We can make constructors public later without it being a breaking change, but we can't take them away, so I'm still leaning towards being optimistic about record types at this time.

Reply:
OK. It sounds like you've considered all the angles I was thinking of, so I yield to your judgment.